clusterloader2/pkg/measurement/common/kube_state_metrics_measurement.go

```diff
@@ -0,0 +1,221 @@
+	result := &measurementutil.LatencyMetric{}
+	err := SetQuantileFromHistogram(result, finalLatency)
+	if err != nil {
+		return nil, err
+	}
+	content, err := util.PrettyPrintJSON(result)
+	if err != nil {
+		return nil, err
+	}
+	// Create Summary.
+	metricsSummary := measurement.CreateSummary(schedulerLatencyMetricName, "json", content)
+	return []measurement.Summary{metricsSummary}, nil
+}
+
+func (m *ksmLatencyMeasurement) start(client clientset.Interface, probeDuration time.Duration) error {
+	m.isRunning = true
+	m.wg.Add(1)
+	var err error
+	var latHist *measurementutil.Histogram
+	go func() {
+		defer m.wg.Done()
+		for {
+			select {
+			case <-m.stopCh:
+				return
+			case <-time.After(probeDuration):
+				latHist, err = m.retrieveKSMLatencyMetrics(client, probeDuration)
+				m.latency = append(m.latency, ksmLatencyMetrics{
+					latencyMetrics: latHist,
+				})
+			}
+		}
+	}()
+	if err != nil {
+		return err
+	}
+	return err
+
+}
+func (m *ksmLatencyMeasurement) retrieveKSMLatencyMetrics(c clientset.Interface, probeDuration time.Duration) (*measurementutil.Histogram, error) {
+	ksmHist := measurementutil.NewHistogram(nil)
+	output, err := m.getMetricsFromService(c, probeDuration)
+	if err != nil {
+		return ksmHist, err
+	}
+	samples, err := measurementutil.ExtractMetricSamples(output)
+	if err != nil {
+		return ksmHist, err
+	}
+	for _, sample := range samples {
+		switch sample.Metric[model.MetricNameLabel] {
+		case ksmRequestDurationMetricName:
+			measurementutil.ConvertSampleToHistogram(sample, ksmHist)
+		}
+	}
+	return ksmHist, nil
+}
+
+func (m *ksmLatencyMeasurement) getMetricsFromService(client clientset.Interface, probeDuration time.Duration) (string, error) {
+	m.isRunning = true
+	m.stopCh = make(chan struct{})
+	m.wg.Add(1)
+
+	finished := make(chan struct{}, 1)
+	var err error
+	var outputSelfMetrics []byte
+	go func() {
+		defer m.wg.Done()
+		var output []byte
+		// Perform request against the /metrics endpoint
+		// of metrics it serves.
+		// This is because kube-state-metrics performance is purely
+		// based around how well it can return the /metrics data.
+		output, err = client.CoreV1().RESTClient().Get().
+			Resource("services").
+			SubResource("proxy").
+			Namespace(m.namespace).
+			Name(fmt.Sprintf("%v:%v", m.serviceName, m.metricsPort)).
+			Suffix("metrics").
+			Do(context.TODO()).Raw()
+		if output == nil {
+			err = fmt.Errorf("/metrics endpoint of kube-state-metrics returned no data in namespace: %s from service: %s port: %d", m.namespace, m.serviceName, m.metricsPort)
+		}
+		// Get metrics, including latency, from telemetry /metrics endpoint.
+		outputSelfMetrics, err = client.CoreV1().RESTClient().Get().
+			Resource("services").
+			SubResource("proxy").
+			Namespace(m.namespace).
+			Name(fmt.Sprintf("%v:%v", m.serviceName, m.selfPort)).
+			Suffix("metrics").
+			Do(context.TODO()).Raw()
+		finished <- struct{}{}
+	}()
+	select {
+	case <-time.After(probeDuration):
```

Comment:
Hmm - this handling seems a bit risky, because you're effectively not killing that goroutine.
So you may stack forever it it (and accumulate many of those in theory).

Why is this even needed?

Reply:
Yes I copied this over from another file that gather metrics from this directory, but you are right, I see two options here:
1. Pass context with a timeout and cancel the context, this should fix this issue.
2. Or if you think we should switch to collecting only at the start and end of measurement as per your previous comment, we then won't have this problem.

Let me know which option you would prefer here, thank you!

Reply:
I would prefere (1), but we would probably need to update client library to allow passing context to the calls.

If you have time to do that - it would be amazing.